README.md
# oid-php
[![packagist](https://img.shields.io/packagist/v/y0x54a/oid-php)](https://packagist.org/packages/y0x54a/oid-php)
[![Build Status](https://github.com/y0x54a/oid-php/workflows/ci/badge.svg?branch=main)](https://github.com/y0x54a/oid-php/actions)
[![codecov](https://codecov.io/gh/y0x54a/oid-php/branch/main/graph/badge.svg?token=19XTYSRDLQ)](https://codecov.io/gh/y0x54a/oid-php)

## Installing
```sh
composer require y0x54a/oid-php
```

## Example
```php
use Y0x54a\Oid\Oid453;
use Y0x54a\Oid\Oid563;
```

```php
Oid453::create();
// 12-byte

Oid563::create();
// 14-byte
```

```php
$oid = new Oid453('00112233445566778899aabb');

$oid->getId();
// 00112233445566778899aabb

$oid->getTimestamp();
// 1122867

$oid->getRandom();
// 293490554760

$oid->getIndex();
// 10070715

$oid->__toString();
// 00112233445566778899aabb

Oid453::validate($oid);
// true

Oid453::validate($oid->getId());
// true

Oid453::validate('00112233445566778899AABB');
// false

Oid563::validate($oid);
// false
```

```php
$oid1 = Oid453::create(1122867);
$oid2 = Oid453::create($oid1);

$oid1->getId() === $oid2->getId();
// true

$oid1->getTimestamp();
// 1122867

$oid2->getTimestamp();
// 1122867

Oid453::generate(1122867);
// 00112233...
```

```php
$oid = new Oid563('00112233445566778899aabbccdd');

$oid->getId();
// 00112233445566778899aabbccdd

$oid->getTimestamp();
// 287454020

$oid->getRandom();
// 93898580466090

$oid->getIndex();
// 12307677

$oid->__toString();
// 00112233445566778899aabbccdd

Oid563::validate($oid);
// true

Oid563::validate($oid->getId());
// true

Oid563::validate('00112233445566778899AABBCCDD');
// false

Oid453::validate($oid);
// false
```

```php
$oid1 = Oid563::create(287454020);
$oid2 = Oid563::create($oid1);

$oid1->getId() === $oid2->getId();
// true

$oid1->getTimestamp();
// 287454020

$oid2->getTimestamp();
// 287454020

Oid563::generate(287454020);
// 0011223344...
```

## API

- ### OidInterface

  - **Methods**

  - `getId(): string`
  
  - `getTimestamp(): int`

  - `getRandom(): int`
  
  - `getIndex(): int`

  - `__toString(): string`

- ### Oid453

  - **Methods**

  - `__construct(string | OidInterface $id)`

  - `getId(): string`

  - `getTimestamp(): int`

  - `getRandom(): int`

  - `getIndex(): int`

  - `__toString(): string`

  - **Static Methods**

  - `create(string | OidInterface | int $id = null): Oid453`

  - `validate(string | OidInterface $id): bool`

  - `generate(int $timestamp): string`

- ### Oid563

  - **Methods**

  - `__construct(string | OidInterface $id)`

  - `getId(): string`

  - `getTimestamp(): int`

  - `getRandom(): int`

  - `getIndex(): int`

  - `__toString(): string`

  - **Static Methods**

  - `create(string | OidInterface | int $id = null): Oid563`

  - `validate(string | OidInterface $id): bool`

  - `generate(int $timestamp): string`

CHANGELOG.md
# 0.1.0 - 2022-08-15

- Release

test/src/resources.php
<?php return [
  'validOid453Inputs' => [
    ['000000000000000000000000', 0x00000000, 0x0000000000, 0x000000],
    ['00000000ffffffffffffffff', 0x00000000, 0xffffffffff, 0xffffff],
    ['ffffffffffffffffffffffff', 0xffffffff, 0xffffffffff, 0xffffff],
    ['00112233445566778899aabb', 0x00112233, 0x4455667788, 0x99aabb]
  ],
  
  'invalidOid453Inputs' => [
    '0000000000000000000000',
    'ffffffffffffffffffffffffff',
    'FF0000000000000000000000',
    'ffffffffffffffffffffffFF',
    'aabbccddeeffgghhiijjkkll',
    '00112233445566778899aabbccdd'
  ],

  'validOid563Inputs' => [
    ['0000000000000000000000000000', 0x0000000000, 0x000000000000, 0x000000],
    ['0000000000ffffffffffffffffff', 0x0000000000, 0xffffffffffff, 0xffffff],
    ['ffffffffffffffffffffffffffff', 0xffffffffff, 0xffffffffffff, 0xffffff],
    ['00112233445566778899aabbccdd', 0x0011223344, 0x5566778899aa, 0xbbccdd]
  ],

  'invalidOid563Inputs' => [
    '00000000000000000000000000',
    'ffffffffffffffffffffffffffffff',
    'FF00000000000000000000000000',
    'ffffffffffffffffffffffffffFF',
    'aabbccddeeffgghhiijjkkllmmnn',
    '00112233445566778899aabb'
  ]
];

test/src/Oid453Test.php
<?php declare(strict_types=1);

namespace Y0x54aTest\Oid;

use Throwable;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;
use Y0x54a\Oid\Oid453;

class Oid453Test extends TestCase
{
  protected $resources;

  protected function setUp(): void{
    $this->resources = require(__DIR__.'/resources.php');
  }

  public function testCreate(){
    $t = time();
    $oid = Oid453::create();
    $this->assertGreaterThanOrEqual($oid->getTimestamp(), $t);
  }

  public function testOutOfTimestamp(){
    $oid = Oid453::create(0xffffffff + 1);
    $this->assertEquals($oid->getTimestamp(), 0);
  }

  public function testValidInputs(){
    foreach ($this->resources['validOid453Inputs'] as $input){
      $oid1 = Oid453::create($input[0]);
      $this->assertEquals($oid1->getId(), $input[0]);
      $this->assertEquals($oid1->getTimestamp(), $input[1]);
      $this->assertEquals($oid1->getRandom(), $input[2]);
      $this->assertEquals($oid1->getIndex(), $input[3]);
      $this->assertEquals($oid1->__toString(), $input[0]);

      $oid2 = Oid453::create($oid1);
      $this->assertEquals($oid2->getId(), $oid1->getId());
      $this->assertEquals($oid2->getTimestamp(), $oid1->getTimestamp());
      $this->assertEquals($oid2->getRandom(), $oid1->getRandom());
      $this->assertEquals($oid2->getIndex(), $oid1->getIndex());
      $this->assertEquals($oid2->__toString(), $oid1->__toString());

      $oid3 = Oid453::create($input[1]);
      $this->assertEquals($oid3->getTimestamp(), $input[1]);
    }
  }

  public function testInvalidInputs(){
    foreach ($this->resources['invalidOid453Inputs'] as $input){
      try {
        Oid453::create($input);
      } catch (Throwable $e) {
        $this->assertInstanceOf(InvalidArgumentException::class, $e);
        continue;
      }
      $this->assertTrue(false);
    }
  }
}

test/src/Oid563Test.php
<?php declare(strict_types=1);

namespace Y0x54aTest\Oid;

use Throwable;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;
use Y0x54a\Oid\Oid563;

class Oid563Test extends TestCase
{
  protected $resources;

  protected function setUp(): void{
    $this->resources = require(__DIR__.'/resources.php');
  }

  public function testCreate(){
    $t = time();
    $oid = Oid563::create();
    $this->assertGreaterThanOrEqual($oid->getTimestamp(), $t);
  }

  public function testOutOfTimestamp(){
    $oid = Oid563::create(0xffffffffff + 1);
    $this->assertEquals($oid->getTimestamp(), 0);
  }

  public function testValidInputs(){
    foreach ($this->resources['validOid563Inputs'] as $input){
      $oid1 = Oid563::create($input[0]);
      $this->assertEquals($oid1->getId(), $input[0]);
      $this->assertEquals($oid1->getTimestamp(), $input[1]);
      $this->assertEquals($oid1->getRandom(), $input[2]);
      $this->assertEquals($oid1->getIndex(), $input[3]);
      $this->assertEquals($oid1->__toString(), $input[0]);

      $oid2 = Oid563::create($oid1);
      $this->assertEquals($oid2->getId(), $oid1->getId());
      $this->assertEquals($oid2->getTimestamp(), $oid1->getTimestamp());
      $this->assertEquals($oid2->getRandom(), $oid1->getRandom());
      $this->assertEquals($oid2->getIndex(), $oid1->getIndex());
      $this->assertEquals($oid2->__toString(), $oid1->__toString());

      $oid3 = Oid563::create($input[1]);
      $this->assertEquals($oid3->getTimestamp(), $input[1]);
    }
  }

  public function testInvalidInputs(){
    foreach ($this->resources['invalidOid563Inputs'] as $input){
      try {
        Oid563::create($input);
      } catch (Throwable $e) {
        $this->assertInstanceOf(InvalidArgumentException::class, $e);
        continue;
      }
      $this->assertTrue(false);
    }
  }
}

src/Oid453.php
<?php declare(strict_types=1);

namespace Y0x54a\Oid;

class Oid453 extends AbstractOid
{
  /**
   * @var array
   */
  protected static $generatedRandom;

  /**
   * @var int
   */
  protected static $generatedIndex;

  /**
   * @param string|OidInterface $id
   * @return bool
   */
  public static function validate($id): bool{
    if ($id instanceof OidInterface) {
      $id = $id->getId();
    }
    return (bool)preg_match('/^[0-9a-f]{24}$/', $id);
  }

  /**
   * @param int $timestamp
   * @return string
   */
  public static function generate(int $timestamp): string{
    $id = '';
    $random = static::generateRandom();
    $index = static::generateIndex();
    /* 4-byte time */
    $id .= static::$hexmap[($timestamp >> 24) & 0xff];
    $id .= static::$hexmap[($timestamp >> 16) & 0xff];
    $id .= static::$hexmap[($timestamp >> 8) & 0xff];
    $id .= static::$hexmap[$timestamp & 0xff];
    /* 5-byte random */
    $id .= static::$hexmap[$random[1]];
    $id .= static::$hexmap[$random[2]];
    $id .= static::$hexmap[$random[3]];
    $id .= static::$hexmap[$random[4]];
    $id .= static::$hexmap[$random[5]];
    /* 3-byte index */
    $id .= static::$hexmap[($index >> 16) & 0xff];
    $id .= static::$hexmap[($index >> 8) & 0xff];
    $id .= static::$hexmap[$index & 0xff];
    return $id;
  }

  /**
   * @return array
   */
  protected static function generateRandom(): array{
    if (static::$generatedRandom === null) {
      static::$generatedRandom = unpack('C*', random_bytes(5));
    }
    return static::$generatedRandom;
  }

  /**
   * @return int
   */
  protected static function generateIndex(): int{
    if (static::$generatedIndex === null) {
      static::$generatedIndex = random_int(0, 0xffffff);
    }
    return static::$generatedIndex = (static::$generatedIndex + 1) % 0xffffff;
  }
}

src/Oid563.php
<?php declare(strict_types=1);

namespace Y0x54a\Oid;

class Oid563 extends AbstractOid
{
  /**
   * @var array
   */
  protected static $generatedRandom;

  /**
   * @var int
   */
  protected static $generatedIndex;

  /**
   * @param string|OidInterface $id
   * @return bool
   */
  public static function validate($id): bool{
    if ($id instanceof OidInterface) {
      $id = $id->getId();
    }
    return (bool)preg_match('/^[0-9a-f]{28}$/', $id);
  }

  /**
   * @param int $timestamp
   * @return string
   */
  public static function generate(int $timestamp): string{
    $id = '';
    $random = static::generateRandom();
    $index = static::generateIndex();
    /* 5-byte time */
    $id .= static::$hexmap[(int)($timestamp / (0xffffffff + 1)) & 0xff];
    $id .= static::$hexmap[($timestamp >> 24) & 0xff];
    $id .= static::$hexmap[($timestamp >> 16) & 0xff];
    $id .= static::$hexmap[($timestamp >> 8) & 0xff];
    $id .= static::$hexmap[$timestamp & 0xff];
    /* 6-byte random */
    $id .= static::$hexmap[$random[1]];
    $id .= static::$hexmap[$random[2]];
    $id .= static::$hexmap[$random[3]];
    $id .= static::$hexmap[$random[4]];
    $id .= static::$hexmap[$random[5]];
    $id .= static::$hexmap[$random[6]];
    /* 3-byte index */
    $id .= static::$hexmap[($index >> 16) & 0xff];
    $id .= static::$hexmap[($index >> 8) & 0xff];
    $id .= static::$hexmap[$index & 0xff];
    return $id;
  }

  /**
   * @return array
   */
  protected static function generateRandom(): array{
    if (static::$generatedRandom === null) {
      static::$generatedRandom = unpack('C*', random_bytes(6));
    }
    return static::$generatedRandom;
  }

  /**
   * @return int
   */
  protected static function generateIndex(): int{
    if (static::$generatedIndex === null) {
      static::$generatedIndex = random_int(0, 0xffffff);
    }
    return static::$generatedIndex = (static::$generatedIndex + 1) % 0xffffff;
  }
}

src/AbstractOid.php
<?php declare(strict_types=1);

namespace Y0x54a\Oid;

use InvalidArgumentException;

abstract class AbstractOid implements OidInterface
{
  /**
   * @var string
   */
  protected $id;

  /**
   * @var int
   */
  protected $timestamp;

  /**
   * @var int
   */
  protected $random;

  /**
   * @var int
   */
  protected $index;

  /**
   * @var array
   */
  protected static $hexmap = [
    '00','01','02','03','04','05','06','07','08','09','0a','0b','0c','0d','0e','0f',
    '10','11','12','13','14','15','16','17','18','19','1a','1b','1c','1d','1e','1f',
    '20','21','22','23','24','25','26','27','28','29','2a','2b','2c','2d','2e','2f',
    '30','31','32','33','34','35','36','37','38','39','3a','3b','3c','3d','3e','3f',
    '40','41','42','43','44','45','46','47','48','49','4a','4b','4c','4d','4e','4f',
    '50','51','52','53','54','55','56','57','58','59','5a','5b','5c','5d','5e','5f',
    '60','61','62','63','64','65','66','67','68','69','6a','6b','6c','6d','6e','6f',
    '70','71','72','73','74','75','76','77','78','79','7a','7b','7c','7d','7e','7f',
    '80','81','82','83','84','85','86','87','88','89','8a','8b','8c','8d','8e','8f',
    '90','91','92','93','94','95','96','97','98','99','9a','9b','9c','9d','9e','9f',
    'a0','a1','a2','a3','a4','a5','a6','a7','a8','a9','aa','ab','ac','ad','ae','af',
    'b0','b1','b2','b3','b4','b5','b6','b7','b8','b9','ba','bb','bc','bd','be','bf',
    'c0','c1','c2','c3','c4','c5','c6','c7','c8','c9','ca','cb','cc','cd','ce','cf',
    'd0','d1','d2','d3','d4','d5','d6','d7','d8','d9','da','db','dc','dd','de','df',
    'e0','e1','e2','e3','e4','e5','e6','e7','e8','e9','ea','eb','ec','ed','ee','ef',
    'f0','f1','f2','f3','f4','f5','f6','f7','f8','f9','fa','fb','fc','fd','fe','ff'
  ];

  /**
   * @param string|OidInterface $id
   * @throws InvalidArgumentException
   */
  public function __construct($id) {
    if (!static::validate($id)) {
      throw new InvalidArgumentException('Invalid ID');
    } else if ($id instanceof OidInterface) {
      $id = $id->getId();
    }
    $this->id = $id;
  }

  /**
   * @return string
   */
  public function getId(): string{
    return $this->id;
  }

  /**
   * @return int
   */
  public function getTimestamp(): int{
    if ($this->timestamp === null) {
      $length = strlen($this->id);
      $this->timestamp = hexdec(substr($this->id, 0, $length / 2 - 4));
    }
    return $this->timestamp;
  }

  /**
   * @return int
   */
  public function getRandom(): int{
    if ($this->random === null) {
      $length = strlen($this->id);
      $this->random = hexdec(substr($this->id, $length / 2 - 4, -6));
    }
    return $this->random;
  }

  /**
   * @return int
   */
  public function getIndex(): int{
    if ($this->index === null) {
      $this->index = hexdec(substr($this->id, -6));
    }
    return $this->index;
  }

  /**
   * @return string
   */
  public function __toString(): string{
    return $this->id;
  }

  /**
   * @param string|OidInterface|int|null $id
   * @return OidInterface
   */
  public static function create($id = null): OidInterface{
    if ($id === null) {
      $id = static::generate(time());
    } else if (is_int($id)) {
      $id = static::generate($id);
    }
    return new static($id);
  }

  /**
   * @param string|OidInterface $id
   * @return bool
   */
  abstract public static function validate($id): bool;

  /**
   * @param int $timestamp
   * @return string
   */
  abstract public static function generate(int $timestamp): string;
}